bpf/attach.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpf
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+)
+
+const (
+	BPF_PROG_BINARY_DIR    = "/usr/lib/calico/bpf"
+	ATTACHED_PROG_HASH_DIR = "/var/run/calico/bpf"
+)
+
+type AttachedProgInfo struct {
+	Name string `json:"name"`
+	Hash string `json:"hash"`
+}
+
+func CheckAttachedProgs(iface, progName string) (bool, string, error) {
+	var progInfo AttachedProgInfo
+	var bytesToRead []byte
+
+	binaryName := path.Join(BPF_PROG_BINARY_DIR, progName)
+	hashCmd := exec.Command("sha256sum", binaryName)
+	outBytes, err := hashCmd.Output()
+	if err != nil {
+		return false, "", err
+	}
+	calculatedHash := strings.Split(string(outBytes), " ")[0]
+
+	name := iface + "_" + strings.TrimSuffix(progName, path.Ext(progName)) + ".json"
```

Comment:
I think using the progname in the file won't be quite right here.  Suppose that the BPF log level is changed from `debug` to `info`.  I think you'll end up with both:
* `cali1234_something_debug.json`, and
* `cali1234_something_info.json`

saved on disk.  Then, if the user switches back to `debug`, the code will check the `cali1234_something_debug.json` file and think the right program is attached.

I think the best option would be pass in the name of the hook to this function, say "tc-ingress", "tc-egress", "xdp-ingress" and to save the file as `cali1234_<name-of-hook>.json`.  Then, switching from `debug` to `info` will use the same file.